File: src/catchido/scrapers/weibo.py
import re
import asyncio
import random
from typing import List, Optional, Dict, Any
from loguru import logger
import httpx

from .base import BaseScraper
from ..db.models import MediaItem, MediaType

class WeiboScraper(BaseScraper):
    def __init__(self, cookie: str, config):
        super().__init__(config)
        self.cookie = cookie
        self.mobile_api = "https://m.weibo.cn/api"

    def get_hd_url(self, url: str) -> str:
        """
        Replace thumbnail/mw2048/bmiddle with 'large' (original resolution).
        Example: https://wx1.sinaimg.cn/mw2048/001abc.jpg -> https://wx1.sinaimg.cn/large/001abc.jpg
        """
        if "sinaimg.cn" not in url:
            return url
            
        # Supported paths: thumbnail, mw2048, bmiddle, orwap, square, mw690
        # We replace the size path part
        url_cleaned = re.sub(r'/(thumbnail|mw2048|bmiddle|orj360|woriginal|mw690)/', '/large/', url)
        return url_cleaned

    async def _create_session(self) -> httpx.AsyncClient:
        client = await super()._create_session()
        # Add weibo headers and cookie
        client.headers.update({
            "Referer": "https://m.weibo.cn/",
            "X-Requested-With": "XMLHttpRequest"
        })
        if self.cookie:
            client.headers["Cookie"] = self.cookie
            logger.debug("Weibo cookie loaded into HTTP headers")
        return client

    async def fetch_media(
        self, 
        query_or_username: str, 
        since_id: Optional[str] = None, 
        limit: Optional[int] = None
    ) -> List[MediaItem]:
        """
        Scrape Weibo. If query_or_username is pure digits, treat as Weibo UID.
        Otherwise treat as search query.
        """
        if not self.client:
            await self._create_session()

        if query_or_username.isdigit():
            return await self._fetch_user_posts(query_or_username, since_id, limit)
        else:
            return await self._fetch_search_results(query_or_username, since_id, limit)

    async def _fetch_user_posts(self, user_id: str, since_id: Optional[str], limit: Optional[int]) -> List[MediaItem]:
        logger.info("Fetching Weibo posts for user UID: {}", user_id)
        media_items = []
        page = 1
        has_more = True
        
        # Determine container ID for user posts
        try:
            index_url = f"{self.mobile_api}/container/getIndex"
            params = {"type": "uid", "value": user_id}
            
            await self._throttle()
            resp = await self.client.get(index_url, params=params)
            if resp.status_code != 200:
                logger.error("Weibo index request failed for {}: {}", user_id, resp.status_code)
                return []
                
            index_data = resp.json()
            tabs = index_data.get("data", {}).get("tabsInfo", {}).get("tabs", [])
            container_id = None
            for tab in tabs:
                if tab.get("tab_type") == "weibo":
                    container_id = tab.get("containerid")
                    break
                    
            if not container_id:
                # Fallback to general containerid scheme
                container_id = f"107603{user_id}"

            # Fetch pages
            items_needed = limit or 50
            while len(media_items) < items_needed and has_more:
                # Add delay to avoid block
                delay = self.config.weibo.request_delay + random.uniform(0.5, 1.5)
                await asyncio.sleep(delay)
                
                logger.debug("Fetching Weibo page {} for container {}", page, container_id)
                page_url = f"{self.mobile_api}/container/getIndex"
                page_params = {"containerid": container_id, "page": page}
                
                await self._throttle()
                resp = await self.client.get(page_url, params=page_params)
                if resp.status_code != 200:
                    logger.error("Failed to fetch page {} for Weibo user {}", page, user_id)
                    break
                    
                page_data = resp.json()
                cards = page_data.get("data", {}).get("cards", [])
                
                if not cards:
                    has_more = False
                    break
                    
                new_items = self._parse_cards(cards, since_id)
                if not new_items:
                    # If since_id is hit, we stop scraping further pages
                    if since_id and any(card.get("mblog", {}).get("id") == since_id for card in cards):
                        logger.info("Hit Weibo since_id checkpoint: {}. Stopping.", since_id)
                        has_more = False
                    # Or no media on page
                    if page > 5: # safety break
                        has_more = False
                
                media_items.extend(new_items)
                page += 1
                
            return media_items[:items_needed]
        except Exception as e:
            logger.exception("Error fetching Weibo UID {}: {}", user_id, e)
            return []

    async def _fetch_search_results(self, query: str, since_id: Optional[str], limit: Optional[int]) -> List[MediaItem]:
        logger.info("Searching Weibo for: {}", query)
        if not self.client:
            await self._create_session()
            
        try:
            search_url = f"{self.mobile_api}/container/getIndex"
            params = {
                "containerid": f"100103type=1&q={query}",
                "page_type": "search"
            }
            await self._throttle()
            resp = await self.client.get(search_url, params=params)
            if resp.status_code != 200:
                logger.error("Weibo search request failed: {}", resp.status_code)
                return []
                
            search_data = resp.json()
            cards = search_data.get("data", {}).get("cards", [])
            
            # Find the card containing the search result list
            result_cards = []
            for card in cards:
                if card.get("card_type") == 11: # Card type 11 holds sub-cards
                    result_cards.extend(card.get("card_group", []))
                elif card.get("card_type") == 9: # Card type 9 is direct mblog
                    result_cards.append(card)
                    
            return self._parse_cards(result_cards, since_id)[:(limit or 50)]
        except Exception as e:
            logger.exception("Error searching Weibo for {}: {}", query, e)
            return []

    async def fetch_media_from_url(self, url: str) -> List[MediaItem]:
        """Fetch media from a specific Weibo detail page (e.g. /detail/123456)."""
        # Weibo URLs can be m.weibo.cn/status/49876543210 or weibo.com/12345678/JklmNopq
        # We can extract the post ID from detail/ or status/
        post_id = None
        match = re.search(r'/(detail|status)/([a-zA-Z0-9]+)', url)
        if match:
            post_id = match.group(2)
        else:
            # Try tailing digits
            match_tail = re.search(r'/([0-9]{10,20})', url)
            if match_tail:
                post_id = match_tail.group(1)

        if not post_id:
            logger.error("Could not extract Weibo post ID from URL: {}", url)
            return []
            
        if not self.client:
            await self._create_session()
            
        try:
            detail_url = f"{self.mobile_api}/statuses/show"
            params = {"id": post_id}
            await self._throttle()
            resp = await self.client.get(detail_url, params=params)
            if resp.status_code != 200:
                logger.error("Weibo detail request failed for ID {}: {}", post_id, resp.status_code)
                return []
                
            resp_data = resp.json()
            mblog = resp_data.get("data")
            if not mblog:
                logger.error("Weibo post details not found for ID {}", post_id)
                return []
                
            return self._parse_mblog(mblog)
        except Exception as e:
            logger.exception("Error getting Weibo URL {}: {}", url, e)
            return []

    def _parse_cards(self, cards: List[Dict[str, Any]], since_id: Optional[str] = None) -> List[MediaItem]:
        media_items = []
        for card in cards:
            if card.get("card_type") != 9:
                continue
                
            mblog = card.get("mblog", {})
            post_id = mblog.get("id")
            
            if since_id and post_id == since_id:
                logger.info("Reached since_id checkpoint: {}. Stopping parse.", since_id)
                break
                
            media_items.extend(self._parse_mblog(mblog))
        return media_items

    def _parse_mblog(self, mblog: Dict[str, Any]) -> List[MediaItem]:
        media_items = []
        post_id = mblog.get("id")
        created_at = mblog.get("created_at")
        text = mblog.get("text", "")
        
        # Clean HTML tags from text
        clean_text = re.sub(r'<[^>]+>', '', text)
        
        # Extract hashtags
        hashtags = [tag.strip("#") for tag in re.findall(r'#([^#]+)#', text)]
        
        # Author info
        user_info = mblog.get("user", {})
        author = user_info.get("screen_name", user_info.get("idstr", "unknown"))

        # Image extraction
        pics = mblog.get("pics", [])
        for idx, pic in enumerate(pics):
            large_url = pic.get("large", {}).get("url")
            orig_url = pic.get("url")
            
            if large_url:
                media_items.append(MediaItem(
                    url=self.get_hd_url(large_url),
                    platform="weibo",
                    post_id=post_id,
                    author=author,
                    media_type=MediaType.IMAGE,
                    width=pic.get("large", {}).get("geo", {}).get("width"),
                    height=pic.get("large", {}).get("geo", {}).get("height"),
                    original_url=orig_url,
                    created_at=created_at,
                    text=clean_text,
                    hashtags=hashtags
                ))

        # Video extraction
        # Weibo stores video in page_info structure
        page_info = mblog.get("page_info", {})
        if page_info.get("type") == "video":
            video_url = None
            urls = page_info.get("media_info", {}).get("stream_url_hd") or page_info.get("media_info", {}).get("stream_url")
            if urls:
                video_url = urls
            
            # Alternative urls structure
            if not video_url:
                playback_list = page_info.get("media_info", {}).get("playback_list", [])
                if playback_list:
                    # Get highest definition
                    playback_list.sort(key=lambda x: x.get("play_info", {}).get("bitrate", 0), reverse=True)
                    video_url = playback_list[0].get("play_info", {}).get("play_url")
            
            if video_url:
                media_items.append(MediaItem(
                    url=video_url,
                    platform="weibo",
                    post_id=post_id,
                    author=author,
                    media_type=MediaType.VIDEO,
                    original_url=video_url,
                    created_at=created_at,
                    text=clean_text,
                    hashtags=hashtags
                ))
                
        return media_items

File: src-tauri/src/scrapers/weibo.rs
use regex::Regex;
use serde_json::Value;
use std::time::Duration;
use tokio::time::sleep;

use crate::config::AppConfig;
use super::MediaItem;

pub struct WeiboScraper {
    client: reqwest::Client,
    cookie: String,
    delay: u64,
}

impl WeiboScraper {
    pub fn new(cookie: String, config: &AppConfig) -> Self {
        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert("Referer", reqwest::header::HeaderValue::from_static("https://m.weibo.cn/"));
        headers.insert("X-Requested-With", reqwest::header::HeaderValue::from_static("XMLHttpRequest"));
        headers.insert(
            "User-Agent",
            reqwest::header::HeaderValue::from_static(
                "Mozilla/5.0 (Linux; Android 10; SM-G960F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Mobile Safari/537.36",
            ),
        );
        
        if !cookie.is_empty() {
            if let Ok(cookie_val) = reqwest::header::HeaderValue::from_str(&cookie) {
                headers.insert("Cookie", cookie_val);
            }
        }

        let client = reqwest::Client::builder()
            .default_headers(headers)
            .timeout(Duration::from_seconds(15))
            .build()
            .unwrap_or_default();

        Self {
            client,
            cookie,
            delay: 2, // 2 seconds default delay
        }
    }

    pub fn get_hd_url(&self, url: &str) -> String {
        if !url.contains("sinaimg.cn") {
            return url.to_string();
        }
        let re = Regex::new(r"/(thumbnail|mw2048|bmiddle|orj360|woriginal|mw690)/").unwrap();
        re.replace(url, "/large/").to_string()
    }

    pub async fn fetch_media(
        &self,
        query_or_username: &str,
        since_id: Option<&str>,
        limit: Option<usize>,
    ) -> Result<Vec<MediaItem>, String> {
        if query_or_username.chars().all(|c| c.is_ascii_digit()) {
            self.fetch_user_posts(query_or_username, since_id, limit).await
        } else {
            self.fetch_search_results(query_or_username, since_id, limit).await
        }
    }

    async fn fetch_user_posts(
        &self,
        user_id: &str,
        since_id: Option<&str>,
        limit: Option<usize>,
    ) -> Result<Vec<MediaItem>, String> {
        let mut media_items = Vec::new();
        let limit_val = limit.unwrap_or(50);
        
        // 1. Get containerid
        let index_url = format!("https://m.weibo.cn/api/container/getIndex?type=uid&value={}", user_id);
        let resp = self.client.get(&index_url).send().await
            .map_err(|e| format!("Weibo index request failed: {}", e))?;
            
        let index_json: Value = resp.json().await
            .map_err(|e| format!("Weibo index JSON parse failed: {}", e))?;
            
        let mut container_id = None;
        if let Some(tabs) = index_json.pointer("/data/tabsInfo/tabs").and_then(|t| t.as_array()) {
            for tab in tabs {
                if tab.get("tab_type").and_then(|t| t.as_str()) == Some("weibo") {
                    container_id = tab.get("containerid").and_then(|c| c.as_str()).map(|s| s.to_string());
                    break;
                }
            }
        }
        
        let container_id = container_id.unwrap_or_else(|| format!("107603{}", user_id));
        let mut page = 1;
        let mut has_more = true;

        while media_items.len() < limit_val && has_more {
            if page > 1 {
                sleep(Duration::from_millis(1500)).await;
            }
            
            let page_url = format!(
                "https://m.weibo.cn/api/container/getIndex?containerid={}&page={}",
                container_id, page
            );
            
            let resp = match self.client.get(&page_url).send().await {
                Ok(r) => r,
                Err(e) => {
                    log::warn!("Weibo page {} request failed: {}", page, e);
                    break;
                }
            };
            
            let page_json: Value = match resp.json().await {
                Ok(j) => j,
                Err(e) => {
                    log::warn!("Weibo page {} JSON parse failed: {}", page, e);
                    break;
                }
            };

            let cards = match page_json.pointer("/data/cards").and_then(|c| c.as_array()) {
                Some(c) => c,
                None => {
                    has_more = false;
                    break;
                }
            };

            if cards.is_empty() {
                has_more = false;
                break;
            }

            let mut reached_checkpoint = false;
            for card in cards {
                if card.get("card_type").and_then(|t| t.as_i64()) != Some(9) {
                    continue; // Type 9 is weibo post card
                }
                
                let mblog = match card.get("mblog") {
                    Some(m) => m,
                    None => continue,
                };
                
                let post_id = mblog.get("id").and_then(|i| i.as_str()).unwrap_or("").to_string();
                if post_id.is_empty() {
                    continue;
                }

                if let Some(sid) = since_id {
                    if post_id == sid {
                        log::info!("Reached Weibo checkpoint: {}. Stopping.", sid);
                        reached_checkpoint = true;
                        break;
                    }
                }

                let author = mblog.pointer("/user/screen_name").and_then(|s| s.as_str()).unwrap_or("").to_string();
                let text = mblog.get("text").and_then(|t| t.as_str()).unwrap_or("").to_string();
                
                // Parse pics
                if let Some(pics) = mblog.get("pics").and_then(|p| p.as_array()) {
                    for pic in pics {
                        if let Some(url_str) = pic.pointer("/large/url").or_else(|| pic.get("url")).and_then(|u| u.as_str()) {
                            media_items.push(MediaItem {
                                url: self.get_hd_url(url_str),
                                platform: "weibo".to_string(),
                                post_id: post_id.clone(),
                                author: author.clone(),
                                media_type: "photo".to_string(),
                                original_url: url_str.to_string(),
                                text: text.clone(),
                            });
                        }
                    }
                }
                
                // Retweet pics
                if let Some(retweet) = mblog.get("retweeted_status") {
                    if let Some(pics) = retweet.get("pics").and_then(|p| p.as_array()) {
                        for pic in pics {
                            if let Some(url_str) = pic.pointer("/large/url").or_else(|| pic.get("url")).and_then(|u| u.as_str()) {
                                media_items.push(MediaItem {
                                    url: self.get_hd_url(url_str),
                                    platform: "weibo".to_string(),
                                    post_id: post_id.clone(),
                                    author: author.clone(),
                                    media_type: "photo".to_string(),
                                    original_url: url_str.to_string(),
                                    text: text.clone(),
                                });
                            }
                        }
                    }
                }
            }

            if reached_checkpoint {
                break;
            }
            
            page += 1;
        }

        Ok(media_items)
    }

    async fn fetch_search_results(
        &self,
        query: &str,
        since_id: Option<&str>,
        limit: Option<usize>,
    ) -> Result<Vec<MediaItem>, String> {
        let mut media_items = Vec::new();
        let limit_val = limit.unwrap_or(50);
        let mut page = 1;
        let mut has_more = true;

        while media_items.len() < limit_val && has_more {
            if page > 1 {
                sleep(Duration::from_millis(1500)).await;
            }
            
            let search_url = format!(
                "https://m.weibo.cn/api/container/getIndex?containerid=100103type%3D1%26q%3D{}&page={}",
                urlencoding::encode(query), page
            );
            
            let resp = match self.client.get(&search_url).send().await {
                Ok(r) => r,
                Err(e) => {
                    log::warn!("Weibo search page {} request failed: {}", page, e);
                    break;
                }
            };
            
            let page_json: Value = match resp.json().await {
                Ok(j) => j,
                Err(e) => {
                    log::warn!("Weibo search page {} JSON parse failed: {}", page, e);
                    break;
                }
            };

            let cards = match page_json.pointer("/data/cards").and_then(|c| c.as_array()) {
                Some(c) => c,
                None => {
                    has_more = false;
                    break;
                }
            };

            if cards.is_empty() {
                has_more = false;
                break;
            }

            let mut reached_checkpoint = false;
            for card in cards {
                // Search result cards can have different nested structures.
                // Usually type 9, or a "card_group" containing cards of type 9
                if let Some(card_group) = card.get("card_group").and_then(|g| g.as_array()) {
                    for group_card in card_group {
                        if group_card.get("card_type").and_then(|t| t.as_i64()) == Some(9) {
                            if let Some(mblog) = group_card.get("mblog") {
                                let post_id = mblog.get("id").and_then(|i| i.as_str()).unwrap_or("").to_string();
                                if post_id.is_empty() {
                                    continue;
                                }

                                if let Some(sid) = since_id {
                                    if post_id == sid {
                                        reached_checkpoint = true;
                                        break;
                                    }
                                }

                                let author = mblog.pointer("/user/screen_name").and_then(|s| s.as_str()).unwrap_or("").to_string();
                                let text = mblog.get("text").and_then(|t| t.as_str()).unwrap_or("").to_string();

                                if let Some(pics) = mblog.get("pics").and_then(|p| p.as_array()) {
                                    for pic in pics {
                                        if let Some(url_str) = pic.pointer("/large/url").or_else(|| pic.get("url")).and_then(|u| u.as_str()) {
                                            media_items.push(MediaItem {
                                                url: self.get_hd_url(url_str),
                                                platform: "weibo".to_string(),
                                                post_id: post_id.clone(),
                                                author: author.clone(),
                                                media_type: "photo".to_string(),
                                                original_url: url_str.to_string(),
                                                text: text.clone(),
                                            });
                                        }
                                    }
                                }
                            }
                        }
                    }
                } else if card.get("card_type").and_then(|t| t.as_i64()) == Some(9) {
                    if let Some(mblog) = card.get("mblog") {
                        let post_id = mblog.get("id").and_then(|i| i.as_str()).unwrap_or("").to_string();
                        if post_id.is_empty() {
                            continue;
                        }

                        if let Some(sid) = since_id {
                            if post_id == sid {
                                reached_checkpoint = true;
                                break;
                            }
                        }

                        let author = mblog.pointer("/user/screen_name").and_then(|s| s.as_str()).unwrap_or("").to_string();
                        let text = mblog.get("text").and_then(|t| t.as_str()).unwrap_or("").to_string();

                        if let Some(pics) = mblog.get("pics").and_then(|p| p.as_array()) {
                            for pic in pics {
                                if let Some(url_str) = pic.pointer("/large/url").or_else(|| pic.get("url")).and_then(|u| u.as_str()) {
                                    media_items.push(MediaItem {
                                        url: self.get_hd_url(url_str),
                                        platform: "weibo".to_string(),
                                        post_id: post_id.clone(),
                                        author: author.clone(),
                                        media_type: "photo".to_string(),
                                        original_url: url_str.to_string(),
                                        text: text.clone(),
                                    });
                                }
                            }
                        }
                    }
                }
                
                if reached_checkpoint {
                    break;
                }
            }

            if reached_checkpoint {
                break;
            }

            page += 1;
        }

        Ok(media_items)
    }
}
